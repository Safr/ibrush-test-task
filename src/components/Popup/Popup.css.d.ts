export const Popup: string;
export const popup: string;
export const Error: string;
export const error: string;
export const SpinnerCenter: string;
export const spinnerCenter: string;
export const orderInfo: string;
export const orderVolumes: string;
export const orderVolumesLeft: string;
export const orderVolumesRight: string;
export const TableContainer: string;
export const tableContainer: string;
export const TableItem: string;
export const tableItem: string;
export const BuyPrice: string;
export const buyPrice: string;
export const Up: string;
export const up: string;
export const Down: string;
export const down: string;
