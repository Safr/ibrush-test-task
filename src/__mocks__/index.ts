export const orderBook = {
  BTC_USD: {
    ask: ['6891', '0.005', '34.455'],
    ask_amount: '2413543.16939829',
    ask_quantity: '190.96765367',
    ask_top: '6887.20157116',
    bid: ['6849.7394927', '0.75', '5137.30461952'],
    bid_amount: '286086.48310063',
    bid_quantity: '11524.76654836',
    bid_top: '6875',
  },
};

export const order = {
  avg: '6875.86448007',
  buy_price: '6859.227881',
  high: '6969.54201841',
  last_trade: '6864.01',
  low: '6800',
  sell_price: '6873.71642794',
  updated: 1528752429,
  vol: '1112.01492053',
  vol_curr: '7632881.53471028',
};

export const currentOrder = {
  BTC_USD: {
    avg: '6875.86448007',
    buy_price: '6859.227881',
    high: '6969.54201841',
    last_trade: '6864.01',
    low: '6800',
    sell_price: '6873.71642794',
    updated: 1528752429,
    vol: '1112.01492053',
    vol_curr: '7632881.53471028',
  },
};

export const orders = { currentOrder };
export const changedOrders = [''];
export const orderBookInfo = [['5897.8242', '0.003', '17.6934726']];

// export const fetchOrders = () => Promise.resolve({
//   orders: {
//     BTC_USD: {
//       avg: '6875.86448007',
//       buy_price: '6859.227881',
//       high: '6969.54201841',
//       last_trade: '6864.01',
//       low: '6800',
//       sell_price: '6873.71642794',
//       updated: 1528752429,
//       vol: '1112.01492053',
//       vol_curr: '7632881.53471028',
//     },
//   },
// });
