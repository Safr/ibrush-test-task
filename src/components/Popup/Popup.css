.Popup {
  width: 95%;
  position: absolute;
  top: 0;
  left: 50%;
  padding: 10px 18px 10px 22px;
  overflow: auto;
  transform: translate(-50%, -50%);
  background-color: #fff;
  text-align: right;
  font-size: 14px;
  color: #222;
  border: solid 2px #41403E;
  /* border-radius: 15px 0 0 15px; */
  border-radius: 12px;
}

.Popup::-webkit-scrollbar {
  width: 12px;
}

/* Track */
.Popup::-webkit-scrollbar-track {
  box-shadow: inset 0 0 5px #41403e; 
  /* border: solid 2px #41403E; */
  border-radius: 12px;
  position: absolute;
  z-index: 5;
  /* border-radius: 5px; */
}

/* Handle */
::-webkit-scrollbar-thumb {
  background-color: #bbe5f1; 
  border-radius: 12px;
}

/* Handle on hover */
.Popup::-webkit-scrollbar-thumb:hover {
  background-color: #67c5df; 
}

.Popup h2 {
  text-align: center;
  margin: 0;
  margin-bottom: 20px;
}

.Error {
  font-size: 2rem;
  text-align: center;
}

.SpinnerCenter {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  margin-top: 80px;
  margin-bottom: 50px;
  text-align: center;
}

.SpinnerCenter p {
  margin-bottom: 25px;
}

.Popup a {
  position: absolute;
  top: 7px;
  right: 5px;
  cursor: pointer;
  font-size: 1.4rem;
}

.orderInfo {
  width: 100%;
  text-align: left;
}

.orderVolumes {
  display: flex;
  justify-content: space-between;
  flex-wrap: wrap;
}

.orderVolumesLeft {
  width: 48%;
}

.orderVolumesRight {
  width: 48%;
}

.TableContainer {
  display: flex;
  justify-content: space-between;
}

.TableItem {
  position: relative;
  width: 46%;
  margin-top: 80px;
}

.BuyPrice {
  margin-left: 30px;
}

.BuyPrice.Up::before {
  content: '';
  display: inline-block;
  width: 0;
  height: 0;
  margin-right: 2px;
  border-style: solid;
  border-width: 0 7px 12px 7px;
  border-color: transparent transparent #57cc1d transparent;
}

.Down::before {
  content: '';
  display: inline-block;
  width: 0;
  height: 0;
  margin-right: 2px;
  border-style: solid;
  border-width: 0 7px 12px 7px;
  border-color: transparent transparent #ff0022 transparent;
}

@media screen and (max-width: 768px) {
  .Popup {
    height: 600px;
    padding: 7px 15px 7px 15px;
  }
  .TableContainer {
    flex-direction: column;
    align-items: center;
    text-align: center;
  }

  .TableItem {
    width: 100%;
    margin-top: 50px;
  }
}

@media screen and (max-width: 560px) {
  .Popup {
    height: 500px;
  }
}

@media screen and (max-width: 480px) {
  .Popup {
    height: 400px;
  }
}