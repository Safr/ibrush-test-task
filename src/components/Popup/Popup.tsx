import React from 'react';
import Transition from 'react-transition-group/Transition';
import OrderTable from '../OrderTable/OrderTable';
import Spinner from '../Spinner/Spinner';
import { OrderBookInfo } from '../../types';
import classes from './Popup.css';

const duration = 500;
const defaultStyle = {
  top: 0,
  transition: `all ${duration}ms ease-in-out`,
  opacity: 0,
};

const transitionStyles = {
  entering: { opacity: 0, top: 0 },
  entered:  { opacity: 1, top: '50%' },
  exiting: { opacity: 1, top: '50%' },
  exited:  { opacity: 0, top: -100 },
};

interface Props {
  currentOrder: any | object;
  orderBook: null |object;
  onClosePopup: () => void;
  isOpen: boolean;
  error: string;
}

class Popup extends React.PureComponent<Props> {
  render() {
    const { isOpen, onClosePopup, currentOrder, orderBook, error } = this.props;
    const orderName = currentOrder ? Object.keys(currentOrder)[0] : '';
    let orderBookBuy: OrderBookInfo[];
    let orderBookSell: OrderBookInfo[];
    if (orderBook !== null) {
      orderBookBuy = orderBook[orderName].ask;
      orderBookSell = orderBook[orderName].bid;
    }
    const renderOrderInfo = () => {
      const orderInfo = currentOrder[orderName];
      return (
        <div className={classes.orderInfo}>
          <header>
          <h1>{orderName}
            <span
              className={`${classes.BuyPrice} ${classes[`${currentOrder.className}`]}`}
            >
             {orderInfo.buy_price.replace('.', ',')}
            </span>
          </h1>
          <p>Статистика за 24 часа:</p>
          </header>
          <div className={classes.orderVolumes}>
            <div className={classes.orderVolumesLeft}>
              <p>Объем торгов: <span>{orderInfo.vol}</span></p>
              <p>Объем сделок: <span>{orderInfo.vol_curr}</span></p>
            </div>
            <div className={classes.orderVolumesRight}>
              <p>Средняя цена сделки: <span>{orderInfo.avg}</span></p>
              <p>Максимальная цена сделки: <span>{orderInfo.high}</span></p>
              <p>Минимальная цена сделки: <span>{orderInfo.low}</span></p>
            </div>
          </div>
          {
            !orderBook ?
            <div className={classes.SpinnerCenter}>
              <p>Ждем получения ордеров по валютной паре</p>
              <Spinner />
            </div>
            :
            <div className={classes.TableContainer}>
              <div className={classes.TableItem}>
                <h2>Покупка</h2>
                <OrderTable orderBook={orderBookBuy} />
              </div>
              <div className={classes.TableItem}>
                <h2>Продажа</h2>
                <OrderTable orderBook={orderBookSell} />
              </div>
            </div>}
            {error && <p className={classes.Error}>{error}</p>}
        </div>
      );
    };
    return (
      <Transition in={isOpen} timeout={duration}>
        {(state: string) => (
          <div style={{ ...defaultStyle, ...transitionStyles[state] }} className={`${classes.Popup} ${state}`}>
            {currentOrder && renderOrderInfo()}
            <a onClick={onClosePopup}>&#10005;</a>
          </div>
        )}
    </Transition>
    );
  }
}

export default Popup;
