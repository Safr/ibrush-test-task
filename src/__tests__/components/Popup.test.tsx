import React from 'react';
import { shallow } from 'enzyme';
import Popup from '../../components/Popup/Popup';
import { currentOrder, orderBook } from '../../__mocks__';

describe('Popup Component', () => {
  let wrapper;
  let onClosePopup;
  beforeEach(() => {
    onClosePopup = jest.fn();
    wrapper = shallow(
      <Popup
        currentOrder={currentOrder}
        orderBook={orderBook}
        onClosePopup={onClosePopup}
        isOpen={false}
        error=""
      />,
    );
  });
  test('should render correctly with props', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
