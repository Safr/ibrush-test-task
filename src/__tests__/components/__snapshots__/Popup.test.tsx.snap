// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popup Component should render correctly with props 1`] = `
<Transition
  appear={false}
  enter={true}
  exit={true}
  in={false}
  mountOnEnter={false}
  onEnter={[Function]}
  onEntered={[Function]}
  onEntering={[Function]}
  onExit={[Function]}
  onExited={[Function]}
  onExiting={[Function]}
  timeout={500}
  unmountOnExit={false}
/>
`;
